lib/internal/url.js

```diff
@@ -1453,18 +1455,29 @@ function pathToFileURL(filepath) {
     outURL.hostname = domainToASCII(hostname);
     outURL.pathname = encodePathChars(
       RegExpPrototypeSymbolReplace(backslashRegEx, StringPrototypeSlice(filepath, hostnameEndIndex), '/'));
-  } else {
-    let resolved = path.resolve(filepath);
-    // path.resolve strips trailing slashes so we must add them back
-    const filePathLast = StringPrototypeCharCodeAt(filepath,
-                                                   filepath.length - 1);
-    if ((filePathLast === CHAR_FORWARD_SLASH ||
-         (isWindows && filePathLast === CHAR_BACKWARD_SLASH)) &&
-        resolved[resolved.length - 1] !== path.sep)
-      resolved += '/';
-    outURL.pathname = encodePathChars(resolved);
-  }
-  return outURL;
+    return outURL;
+  }
+  let resolved = path.resolve(filepath);
+  // path.resolve strips trailing slashes so we must add them back
+  const filePathLast = StringPrototypeCharCodeAt(filepath,
+                                                 filepath.length - 1);
+  if ((filePathLast === CHAR_FORWARD_SLASH ||
+       (isWindows && filePathLast === CHAR_BACKWARD_SLASH)) &&
+      resolved[resolved.length - 1] !== path.sep)
+    resolved += '/';
+
+  // Call encodePathChars first to avoid encoding % again for ? and #.
+  resolved = encodePathChars(resolved);
+
+  // Question and hash character should be included in pathname.
+  // Therefore, encoding is required to eliminate parsing them in different states.
+  // This is done as an optimization to not creating a URL instance and
+  // later triggering pathname setter, which impacts performance
+  if (StringPrototypeIncludes(resolved, '?'))
+    resolved = RegExpPrototypeSymbolReplace(questionRegex, resolved, '%3F');
```

Comment:
Ok but can't you use `string.replace('?', '%3F')` instead of `string.replace(/\?/, '%3F')`?

Reply:
> Do you have a benchmark that shows otherwise?

No, `RegExpPrototypeSymbolReplace()` also replaces only the first item with `/\?/` so it seems only a complication.

```
$ node
Welcome to Node.js v20.4.0.
Type ".help" for more information.
> var re = /\?/
undefined
> re[Symbol.replace]('???', 'x')
'x??'
>
```

Reply:
Yes, it seems I forgot to pass `g` at the end.